HIP/hip-584.md

```diff
@@ -0,0 +1,128 @@
+---
+hip: 584
+title: EVM Module
+author:  Ivan Kavaldzhiev <ivan.kavaldzhiev@limechain.tech>, Mustafa Uzun <mustafa.uzun@limechain.tech>
+type: Standards Track
+category: Service
+needs-council-approval: Yes
+status: Review
+created: 2022-09-27
+discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/discussions/586
+updated: 2022-10-04
+---
+
+## Abstract
+
+Encapsulating Hedera smart contract EVM specific logic in a library.
+It would expose interfaces that will allow different implementations.
+Mirror node would build ethereum-related REST API on top of the library. 
+Mirror nodes could use the library to simulate a transaction and the output would be used to 
+populate the return value for the REST API that is invoked. 
+Consensus nodes would also use the library and keep their existing logic by providing specific interface implementations.
+
+## Motivation
+
+Extracting common Hedera smart contract EVM logic in a single place
+
+Enabling free read-only queries.
+
+Enabling free gas estimate for a given transaction.
+
+Enabling free transaction simulations.
+
+## Rationale
+
+In order to be possible to both execute and simulate transactions, run by the EVM used inside `hedera-node`,
+we need to extract a module encapsulating all of the EVM logic.
+
+The resulting library would allow us to fulfil the following use cases:
+
+- Execute transactions in a regular flow with committing changes to state
+- Execute transaction in a regular flow or simulations that perform read-only operations
+- Execute transaction simulations that perform speculative writes without committing changes to state
+- Execute transaction simulations that perform speculative writes using historical data for transaction replay, 
+without committing changes to state
+
+## User stories
+
+- As a developer, I want to have a library with common Hedera smart contract EVM logic that I can reuse in different projects
+- As a user, I want to perform `eth_call` calls from the Ethereum JSON-RPC standard
+- As a user, I want to perform `eth_gasEstimate` calls from the Ethereum JSON-RPC standard
+
```

Comment:
I think we need a story for the debug use case, worded so it's just the enabling logic and not the full debug support

```suggestion
- As a user, I want a library that can be used to re-run historical EVM transactions accurately so that I can inspect and debug the execution of the bytecode logic.

```